### PR TITLE
Tyler/firmware shell

### PR DESCRIPTION
**Pull request by tyhoff**

### CODE REVIEW

File: _drafts/firmware-shell.md [hunk 1]
```diff
@@ -0,0 +1,991 @@
...
+
+The final part would be to add it to the list within `shell_commands.c`.
+
+````c
+static const sShellCommand s_shell_commands[] = {
+  {"kv_write", cli_cmd_kv_write, "Write a Key/Value pair"},  // Added!
+  {"hello", cli_cmd_hello, "Say hello"},
+  {"help", shell_help_handler, "Lists all commands"},
+};
+
```
Comment (franc0is) [inline _drafts/firmware-shell.md:559]:
```suggestion
```
```

**Comment by tyhoff** [comment]
> Great edits. Fix that code section and you're good to go :-)

Thank you for the reviews! Was an extra kick of motivation I needed to do the post right.

File: _drafts/firmware-shell.md [hunk 2]
```diff
@@ -0,0 +1,992 @@
...
+````
+
+Here's what happens when we run this script:
+
+```
+$ python auto_serial.py
+Hello World!
+```
+
+The ability opens up a huge amount of automation opportunities. By using scripts
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:598]:
```suggestion
This ability opens up a huge amount of automation opportunities. By using scripts
```

File: _drafts/firmware-shell.md [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+to interact with the shell, you can easily:
+
+- Read the values of arbitrary variables and regions of RAM or flash for
+  inspection without the use of a debugger.
+- Send and receive files from a device's file system.
+- Run a suite of automated tests, all of which are individual shell commands
+  that return success codes.
+- Perform before and after analysis on heaps, battery life, etc. when running
+  automated tests.
+- Manage a large number of devices connected to a single computer.s
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:608]:
```suggestion
- Manage a large number of devices connected to a single computer.
```

File: _drafts/firmware-shell.md [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+These might seem cryptic at first, but this is the standard way to passing
+command-line arguments to C/C++ programs.[^command_line_args]
+
+**argc** - number of strings pointed to by `argv`<br> **argv** - 1 plus the
+number of arguments passed to the function. The first argument is the command
+name.<br>
+
+To further understand how these two arguments work together in our shell, let's
+build a command called `kv_write` that could hook into the Key/Value Store from
+a previous Interrupt post, [Unit Testing
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:513]: this link doesn't appear to have rendered right

File: _drafts/firmware-shell.md [hunk 5]
```diff
@@ -0,0 +1,992 @@
...
+
+void shell_result_fail(int code) {
+  char buf[32];
+  snprint(buf, sizeof(buf), "> FAIL,%d", code);
+  shell_put_line(buf);
+}
+
+```
+
+One common format that is used in CSV, which is mostly readable depending on
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:770]:
```suggestion
One common format that is used is CSV, which is mostly readable depending on
```

File: _drafts/firmware-shell.md [hunk 6]
```diff
@@ -0,0 +1,992 @@
...
+in-place editing, Emacs shortcuts, command history and completion, argument
+parsing, and help menus.
+
+To learn how to add some of these features to your project's shell, it's best to
+look at a few open-source projects. Unfortunately, there aren't many _great_
+shell's publicly available for tiny firmware.
+
+- [Zephyr Shell](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/shell)
+- [Skip Scooters Anchor Console](https://github.com/rideskip/anchor/tree/master/console)
+
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:800]: The esp-idf shell does some neat stuff too. A lot of it is offloaded to the desktop:  https://github.com/espressif/esp-idf/blob/master/tools/idf_monitor.py#L3-L10

console source itself is here https://github.com/espressif/esp-idf/tree/master/components/console

File: _drafts/firmware-shell.md [hunk 7]
```diff
@@ -0,0 +1,992 @@
...
+- You need to maintain multiple connections to a single device over a single
+  serial port.
+- You are sure that any consumer of the interface (developers, QA, factory
+  harnesses) will use the tools your team builds. In this case, I would
+  definitely go for the binary protocol route.
+
+## Final Thoughts
+
+As an evangelist for developer productivity, I love finding repeated tasks and
+automated the processes, whether that is in the firmware shell, a [project CLI
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:963]:
```suggestion
automating the processes, whether that is in the firmware shell, a [project CLI
```

File: _drafts/firmware-shell.md [hunk 8]
```diff
@@ -0,0 +1,992 @@
...
+on the host
+machine]({% post_url 2019-08-27-building-a-cli-for-firmware-projects %}), or in
+[the
+debugger]({% post_url 2019-07-02-automate-debugging-with-gdb-python-api %}).
+
+I hope this post was found to be useful and help tell the story why having a
+developer-focused interface into a firmware-based device is helpful. I'd love to
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:970]: maybe: sentence seems a little wordy, perhaps drop first part
```suggestion
I hope this post conveyed the reasons why having a developer-focused interface into a firmware-based device is helpful. I'd love to
```

File: _drafts/firmware-shell.md [hunk 9]
```diff
@@ -0,0 +1,992 @@
+---
+title: Building a Tiny CLI Shell for Tiny Firmware
+description:
```
Comment (chrisc11) [inline _drafts/firmware-shell.md:3]: just noting, still need a description (and thumbnail)

**Pull request merged**